block-explorer/backend/src/api_v0/mod.rs

```diff
@@ -0,0 +1,506 @@
+        ),
+        links: links.links(),
+        errors: vec![],
+    };
+    (StatusCode::OK, Json(response_obj))
+}
+
+async fn get_events(
+    State(state): AxumState,
+    ValidatedQuery(params): ValidatedQuery<models::EventsQuery>,
+    OriginalUri(uri): OriginalUri,
+) -> (StatusCode, Json<ResponseObject<models::Event>>) {
+    let events = match state.db.get_events(&params).await {
+        Ok(events) => events,
+        Err(err) => return gateway_timeout_response_504(err),
+    };
+
+    let full_url = format!("{}{}", state.base_url, uri);
+    let mut links = Links::new(state.base_url.clone());
+    links.add("self", uri.to_string());
+    links.add_pagination_links(&params.pagination.links(&full_url, "FIXME"));
+
+    let response_obj = ResponseObject {
+        data: Some(ResponseObjectData::Many(
+            events
+                .into_iter()
+                .map(|event| ResourceObject {
+                    resoure_type: "event",
+                    id: event.id.to_string(),
+                    attributes: event,
+                })
+                .collect(),
+        )),
+        errors: vec![],
+        links: links.links(),
+    };
+    (StatusCode::OK, Json(response_obj))
+}
+
+async fn get_blocks(
+    State(state): AxumState,
+    Query(params): Query<models::BlocksQuery>,
+    OriginalUri(uri): OriginalUri,
+) -> (StatusCode, Json<ResponseObject<JsonValue>>) {
+    let blocks = match state.db.get_blocks(&params).await {
+        Ok(blocks) => blocks,
+        Err(err) => {
+            return gateway_timeout_response_504(err);
+        }
+    };
+
+    let full_url = format!("{}{}", state.base_url, uri);
+    let mut links = Links::new(state.base_url.clone());
+    links.add("self", uri.to_string());
+    links.add_pagination_links(&params.pagination.links(&full_url, "FIXME"));
+
+    let response_obj = ResponseObject {
+        data: Some(ResponseObjectData::Many(
+            blocks
+                .into_iter()
+                .map(|block| ResourceObject {
+                    resoure_type: "block",
+                    id: block["hash"].as_str().unwrap().to_string(),
+                    attributes: block,
+                })
+                .collect(),
+        )),
+        errors: vec![],
+        links: links.links(),
+    };
+    (StatusCode::OK, Json(response_obj))
+}
+
+async fn get_transactions(
+    State(state): AxumState,
+    ValidatedQuery(params): ValidatedQuery<models::TransactionsQuery>,
+    OriginalUri(uri): OriginalUri,
+) -> (StatusCode, Json<ResponseObject<JsonValue>>) {
+    let txs = match state.db.get_transactions(&params).await {
+        Ok(txs) => txs,
+        Err(err) => {
+            return gateway_timeout_response_504(err);
+        }
+    };
+
+    let full_url = format!("{}{}", state.base_url, uri);
+    let mut links = Links::new(state.base_url.clone());
+    links.add("self", uri.to_string());
+    links.add_pagination_links(&params.pagination.links(&full_url, "FIXME"));
```

Comment:
The above 4 line are repeated in `get_transactions , get_blocks and get_events` should the methods somehow share it?

Reply:
Indeed! Fixed.